### PR TITLE
1408 Upload and Download Functional Use

### PR DESCRIPTION
**Pull request by mwfrost**
Closes: #1408 
This branch adds the ability to download a csv table of all the reported functional use records associated with a data group. The user can then automatically derive clean_funcuse values for each record based on the report_funcuse. A new csv file with the assigned values can be uploaded, and the clean_funcuse of each existing record will be updated to reflect the csv file's contents.

### CODE REVIEW

File: dashboard/fixtures/18_functional_use.yaml [hunk 1]
```diff
@@ -122,3 +122,24 @@
     chem: 854
     report_funcuse: surfactaint
     clean_funcuse: ""
+- model: dashboard.functionaluse
+  pk: 13
+  fields:
+    created_at: 2020-01-30 02:45:44.193517
+    updated_at: 2020-01-30 02:45:44.193532
+    chem: 86
+    report_funcuse: fresh scennt
```
Comment (BryanMG957) [inline dashboard/fixtures/18_functional_use.yaml:131]: Is "fresh sceNNt" a typo? @mwfrost

Comment (mwfrost) [reply dashboard/fixtures/18_functional_use.yaml:131]: @BryanMG957 Yes - the idea is that the misspelled word was in `report_funcuse` and the cleaning process would convert it to "fresh scent"

File: dashboard/views/data_group.py [hunk 2]
```diff
@@ -416,3 +449,27 @@ def get_product_csv_template(request, pk):
...
+    ]
+    qs = FunctionalUse.objects.filter(
+        chem__extracted_text__data_document__data_group=datagroup
+    ).values(*columnlist)
+    return render_to_csv_response(
+        qs,
+        filename=(datagroup.get_name_as_slug() + "_raw_functional_use_records.csv"),
+        field_header_map={
+            "extraction_script__title": "extraction_script",
+            "category__title": "category",
```
Comment (michael-barbour) [inline dashboard/views/data_group.py:472]: On the upload this header is called category_title.  These should probably be the same, either by changing the upload header or this one.

File: dashboard/forms/data_group.py [hunk 3]
```diff
@@ -772,3 +773,69 @@ class Meta:
...
+            raise forms.ValidationError(f"CSV column titles should be {header_cols}")
+
+        for f in self.forms:
+            # Convert category title string to an existing functional use category object
+            category_title = f.cleaned_data.get("category_title")
+            category = FunctionalUseCategory.objects.filter(
+                title=category_title
+            ).first()
+            if not category:
+                raise forms.ValidationError(f"Invalid category title provided.")
```
[the remark this answers is not in the document]
Comment (mwfrost) [reply dashboard/forms/data_group.py:821]: Great idea. I have added the invalid category to the row's error message.

File: dashboard/forms/data_group.py [hunk 4]
```diff
@@ -772,3 +773,69 @@ class Meta:
...
+class FunctionalUseCSVForm(forms.Form):
+    # A ModelForm does not work here because it will
+    # not carry the id field
+    id = forms.IntegerField(required=True)
+    report_funcuse = field_for_model(FunctionalUse, "report_funcuse")
+    category_title = forms.CharField(required=True)
+    clean_funcuse = field_for_model(FunctionalUse, "clean_funcuse")
+
+    class Meta:
+        fields = ["id", "report_funcuse", "category_title", "clean_funcuse"]
```
[the remark this answers is not in the document]
Comment (mwfrost) [reply dashboard/forms/data_group.py:787]: removed this from the upload (but marking this spot for when the users realize they actually wanted that field in the upload)

File: dashboard/views/data_group.py (hunk 2 above)
Comment (mwfrost) [reply dashboard/views/data_group.py]: Fixed.

File: templates/data_group/datagroup_detail.html [hunk 5]
```diff
@@ -124,6 +124,9 @@ <h1>{{ datagroup.name }}</h1>
               <a class="nav-link" id="extfile_formset-nav-link" data-toggle="tab" role="tab" href="#extfile_formset">Extracted</a>
             </li>
             {% elif datagroup.type == 'FU' or datagroup.type == 'CO' or datagroup.type == 'CP' %}
+            <li class="nav-item">
+                 <a class="nav-link" id="functional_uses_upload_form-nav-link" data-toggle="tab" role="tab" href="#functional_uses_upload_form">Func. Use</a>
```
[the remark this answers is not in the document]
Comment (mwfrost) [reply templates/data_group/datagroup_detail.html:128]: I don't know how serious this is. I'll ask Kathie.

[the remark this answers is not in the document]
Comment (mwfrost) [reply templates/data_group/datagroup_detail.html:128]: They have meetings today - let's call it good for now and if Kathie wants a different layout in that panel, we can fix it in a new ticket

**Pull request merged**